pkg/instance/service.go

```diff
@@ -76,43 +80,224 @@ func (s service) SaveInstance(instance *model.DeploymentInstance) error {
 	}
 
 	deployment.Instances = append(deployment.Instances, instance)
-	err = s.validateDeployment(deployment)
+
+	_, err = s.validateNoCycles(deployment.Instances)
+	if err != nil {
+		return err
+	}
+
+	err = s.resolveParameters(deployment)
 	if err != nil {
 		return err
 	}
 
 	return s.instanceRepository.SaveInstance(instance)
 }
 
-func (s service) validateDeployment(deployment *model.Deployment) error {
-	stacks, err := s.collectStacks(deployment)
+func (s service) validateNoCycles(instances []*model.DeploymentInstance) (graph.Graph[string, *model.DeploymentInstance], error) {
+	g := graph.New(func(instance *model.DeploymentInstance) string {
+		return instance.StackName
+	}, graph.Directed(), graph.PreventCycles())
+
+	for _, instance := range instances {
+		err := g.AddVertex(instance)
+		if err != nil {
+			if errors.Is(err, graph.ErrVertexAlreadyExists) {
+				return nil, fmt.Errorf("failed adding instance for stack %q as one already exists", instance.StackName)
+			}
+			return nil, fmt.Errorf("failed adding vertex for instance %q: %v", instance.Name, err)
+		}
+	}
+
+	for _, src := range instances {
+		stack, err := s.stackService.Find(src.StackName)
+		if err != nil {
+			return nil, err
+		}
+		for _, requiredStack := range stack.Requires {
+			err := g.AddEdge(src.StackName, requiredStack.Name)
+			if err != nil {
+				if errors.Is(err, graph.ErrEdgeAlreadyExists) {
+					return nil, fmt.Errorf("instance %q requires %q more than once", src.Name, requiredStack.Name)
+				} else if errors.Is(err, graph.ErrEdgeCreatesCycle) {
+					return nil, fmt.Errorf("link from instance %q to instance %q creates a cycle", src.Name, requiredStack.Name)
+				}
+				return nil, fmt.Errorf("failed linking instance %q with instance %q: %v", src.Name, requiredStack.Name, err)
+			}
+		}
+	}
+
+	return g, nil
+}
+
+func (s service) resolveParameters(deployment *model.Deployment) error {
+	for _, instance := range deployment.Instances {
+		stack, err := s.stackService.Find(instance.StackName)
+		if err != nil {
+			return err
+		}
+
+		instanceParameters := instance.Parameters
+		err = rejectNonExistingParameters(instanceParameters, stack)
+		if err != nil {
+			return err
+		}
+
+		err = rejectConsumedParameters(instanceParameters, stack)
+		if err != nil {
+			return err
+		}
+
+		addDefaultParameterValues(stack, instanceParameters)
```

Comment:
Please switch the parameters so we have a consistent order of `instanceParameters, stack` in all the methods we call here.